tensorflow/compiler/tf2tensorrt/utils/trt_tensor_proxy.h

```diff
@@ -0,0 +1,599 @@
+    }
+    assert(0 && "Should not come here.");   
+  }
+
+  void setAllowedFormats(nvinfer1::TensorFormats formats)
+  {
+    switch(ttype_)
+    {
+      case TensorType::kTRT: return trt_tensor_->setAllowedFormats(formats);
+      case TensorType::kSIMPLE: return simple_tensor_->setAllowedFormats(formats);
+      case TensorType::kFAKE: return fake_tensor_->setAllowedFormats(formats);
+      case TensorType::kNONE: assert(0 && "Unsupported itensor_ type");
+    }
+    assert(0 && "Should not come here.");   
+  }
+
+  nvinfer1::TensorFormats getAllowedFormats() const
+  {
+    switch(ttype_)
+    {
+      case TensorType::kTRT: return trt_tensor_->getAllowedFormats();
+      case TensorType::kSIMPLE: return simple_tensor_->getAllowedFormats();
+      case TensorType::kFAKE: return fake_tensor_->getAllowedFormats();
+      case TensorType::kNONE: assert(0 && "Unsupported itensor_ type");
+    }
+    assert(0 && "Should not come here.");   
+  }
+
+  bool isShapeTensor() const
+  {
+    switch(ttype_)
+    {
+      case TensorType::kTRT: return trt_tensor_->isShapeTensor();
+      case TensorType::kSIMPLE: return simple_tensor_->isShapeTensor();
+      case TensorType::kFAKE: return fake_tensor_->isShapeTensor();
+      case TensorType::kNONE: assert(0 && "Unsupported itensor_ type");
+    }
+    assert(0 && "Should not come here.");   
+  }
+
+  bool isExecutionTensor() const
+  {
+    switch(ttype_)
+    {
+      case TensorType::kTRT: return trt_tensor_->isExecutionTensor();
+      case TensorType::kSIMPLE: return simple_tensor_->isExecutionTensor();
+      case TensorType::kFAKE: return fake_tensor_->isExecutionTensor();
+      case TensorType::kNONE: assert(0 && "Unsupported itensor_ type");
+    }
+    assert(0 && "Should not come here.");   
+  }
+
+private:
+  bool validate() const
+  {
+    return trt_tensor_ && !simple_tensor_ && !fake_tensor_
+      || !trt_tensor_ && simple_tensor_ && !fake_tensor_
+      || !trt_tensor_ && !simple_tensor_ && fake_tensor_;
+  }
+
+  TensorType ttype_{TensorType::kNONE};
+
+  // When ITensorProxy represents an ITensor, the ITensor can be either passed
+  // by the caller via the constructor that takes an ITensor* as parameter, or
+  // be created as a SimpleITensor.
+  //
+  // In the first case, the ITensor pointer is stored in 'tensor_' below, and
+  // the ITensor itself is not owned by this class. This method is used by
+  // Converter (e.g. AddInputTensor) and op converters during TRT network
+  // construction, where the TRT network owns the ITensor.
+  //
+  nvinfer1::ITensor* trt_tensor_ = nullptr; // Not owned.
+  // In the second case, the created SimpleITensor is stored in
+  // 'simple_itensor_' below and is owned by this class. SimpleITensor is a fake
+  // implementation of ITensor and is used only by TrtNodeValidator to validate
+  // the graph nodes.
+  std::shared_ptr<SimpleITensor> simple_tensor_ = nullptr;
+  std::unique_ptr<FakeITensor> fake_tensor_ = nullptr;
```

Comment:
Shouldn't they be in a union?

Reply:
This is done with the class merge, thanks